File: python/509FibonacciNumber.py
class Solution:
    def getFib(self,n,d) -> int:
        if d.get(n) is not None:
            return d.get(n)
        else:
            if n == 0 or n == 1:
                d[n] = n
                return d[n]
            else:
                d[n] = self.getFib(n-1,d) + self.getFib(n-2,d)
                return d[n]

    def fib(self, n: int) -> int:
        a=dict();
        return self.getFib(n,a)

File: rust/509FibonacciNumber.rs
impl Solution {
	pub fn fib(n: i32) -> i32 {
			let mut a = 0;
			let mut b = 1;
			// 0,1 when n = 2, i.e. 2 iterations
			for _ in 0..n {
					let temp = b;
					b = a + b;
					a = temp;
			}
			return a;
	}
}

impl Solution {
	pub fn fib(n: i32) -> i32 {
			if n == 0 {
					return 0;
			}
			if  n == 1 {
					return 1;
			}
			return Self::fib(n-1) + Self::fib(n - 2);
	}
}

impl Solution {
	pub fn fib(n: i32) -> i32 {
			(0..n)
					.fold((0, 1), |(a, b), _| (b, a + b))
					.0
	}
}

impl Solution {
	pub fn fib(n: i32) -> i32 {
			fn calculate (n: i32) -> i32 {
					if n == 1 {
							return 1;
					}
					if n == 0 {
							return 0;
					}
					return calculate(n - 1) + calculate(n - 2);
			}

			calculate(n)
	}
}	